flink-core/src/main/java/org/apache/flink/core/plugin/DirectoryBasedPluginFinder.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.plugin;
+
+import org.apache.flink.util.function.FunctionUtils;
+
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to create a collection of {@link PluginDescriptor} based on directory structure for a given plugin
+ * root folder.
+ *
+ * <p>The expected structure is as follows: the given plugins root folder, containing the plugins folder. One plugin folder
+ * contains all resources (jar files) belonging to a plugin. The name of the plugin folder becomes the plugin id.
+ * <pre>
+ * plugins-root-folder/
+ *            |------------plugin-a/ (folder of plugin a)
+ *            |                |-plugin-a-1.jar (the jars containing the classes of plugin a)
+ *            |                |-plugin-a-2.jar
+ *            |                |-...
+ *            |
+ *            |------------plugin-b/
+ *            |                |-plugin-b-1.jar
+ *           ...               |-...
+ * </pre>
```

Comment:
Would it better to add another layer of plugin type ? Something like this:

plugins/FileSystem/HadoopFileSystem
plugins/FileSystem/S3FileSystem
plugins/Reporter/Reporter1
plugins/Reporter/Reporter2

Reply:
Another concern is that whether this plugin finder mechanism works in JM/TM ?

Reply:
Currently plugins are loaded both on JM and TM.

Regarding the directory structure, the downside of splitting it per type, is increased complexity for a user and extra possibilities for misconfigurations (typos in subdirectories, placing plugin in wrong subdirectory). Also it would prevent from implementing a single plugin that adds multiple things. Think of for example a plugin adding both SQL data type and UDFs to manipulate it at the same time.

The only benefit that I can think of is easier/faster plugin discovery (when looking for file systems you could skip reporter plugins), however I'm not sure if that's an issue. Also if it will prove to be an issue, we can opt for centralised plugin architecture, something like [presto is doing](https://github.com/prestodb/presto/blob/master/presto-spi/src/main/java/com/facebook/presto/spi/Plugin.java) - we could discover only `Plugin` implementations and `Plugin` would report what does it provide or not.

Reply:
For yarn mode, we should ship the plugins to yarn container. I don't think this PR do that, but it could be done in another PR.

Reply:
Yes @zjffdu you are right. There is another [jira tikcet for that](https://issues.apache.org/jira/browse/FLINK-12143).

Reply:
Thanks @pnowojski